### PR TITLE
Use uniquify.el for uniquifiying buffer names

### PR DESCRIPTION
**Pull request by tarsius**
(Re-roll of  #2056) @npostavs said:

There might be a problem if `topdir` is `nil`. I couldn't remember under what scenario that happens.

### CODE REVIEW

**Comment by tarsius** [comment]
My squash commit was created using the new `magit-commit-augment` instead of `magit-commit-squash` which does not allow editing the message. I used that to comment on the issue you raised above.

What's not so nice about this is that you cannot see that comment here, but first have to click on the link above. But in the commit message I have mentioned you - did that cause a Github notification?

**Comment by tarsius** [comment]
Okay, now back to your patch...

This doesn't seem to do anything for me, do I have to do anything to enable this feature?

Also this should be optional, probably depending on what %-sequence is used in the buffer-name format. We could repurpose %b, or replace it with %r.

**Comment by npostavs** [comment]
> But in the commit message I have mentioned you - did that cause a Github notification?

Yup. It showed up in by inbox first.

> This doesn't seem to do anything for me, do I have to do anything to enable this feature?

No it should be turned on from the start. You can use the recipe from #2054 to test it. Um, just realized that only tests magit-status buffers, and I forgot to enable it on other modes.

> Also this should be optional, probably depending on what %-sequence is used in the buffer-name format. We could repurpose %b, or replace it with %r.

Oh, it's already dependent on using `%b` in the format, otherwise there will be no buffer name conflicts, and uniquify.el doesn't do anything.

**Comment by tarsius** [comment]
I have tried this again, and I like it.

The name is uniquified by adding a string *after* the `*`. If it's not feasible to put it before that, then I would be okay with dropping both stars. But first try if you can keep the trailing star and keep it at the end.

I think we could also make it the default like this:

* Drop support for plain "basename" variant. As we have seen it has issues.
* Use `%u` to mean "uniquify".
* Treat `%b` as a synonym for `%u`, for the sake of people who have already customized the formats.
* Default to uniquify by replacing `%a` with `%u` in all formats.
* Add a variable to turn of uniquify globally, so that users don't have to customize each format individually.
* Adjust the documentation. This should also make it into the manual.

**Comment by npostavs** [comment]
> The name is uniquified by adding a string after the `*`. If it's not feasible to put it before that, then I would be okay with dropping both stars. But first try if you can keep the trailing star and keep it at the end.

Okay, I'll investigate this.

> Drop support for plain "basename" variant. As we have seen it has issues.

If you refer to #2054, I believe the issue is the same with uniquify added because uniquify only comes into effect after creating the new buffer.


> Use %u to mean "uniquify".

I think having uniquify acting on %b formats (what's currently pushed here) makes sense. In current magit without uniquify, %b formats get uniquified in the old style `*magit-status: foo*<2>`. Since 24.4, Emacs enables uniquify by default so it makes sense that magit will use `uniquify-buffer-name-style` for this instead.

> Add a variable to turn of uniquify globally, so that users don't have to customize each format individually.

They could use `uniquify-ignore-buffers-re`.

**Comment by npostavs** [comment]
> The name is uniquified by adding a string after the *. If it's not feasible to put it before that, then I would be okay with dropping both stars. But first try if you can keep the trailing star and keep it at the end.

The uniquify code really wants to treat the whole name as a single unit, it doesn't want to put stuff in the middle. It can be worked around by renaming after unquify has done it's job, a bit hacky but it seems to work (b6cf00f23bd7cfe1579f62b82c69c54f648b6508).

> But in the commit message I have mentioned you - did that cause a Github notification?

Fun fact: when I rebased it, I sent myself the notification again.

**Comment by tarsius** [comment]
Alright. Could you please also mention uniquify in the relevant doc-strings. I just noticed that only `magit-status-buffer-name-format` is mentioned in the manual. The other variables should also be documented in the appropriate sections, and, more importantly, the concept should be described somewhere in "Modes and Buffers". If you could do that, that would be great.

Do you, and @kyleam, think we should default to %b?

Also did you investigate the use of uniquify.el when listing known repositories?

**Comment by npostavs** [comment]
> Do you, and @kyleam, think we should default to %b?

I think `%b` with uniquifying is more readable, so yes.

> Also did you investigate the use of uniquify.el when listing known repositories?

No, is this for `magit-list-repos-uniquify` called from `magit-read-repository`? I *suspect* uniquify won't be helpful there because there are no buffers involved, but I'll see if something can be hacked in.

**Comment by kyleam** [comment]
> think we should default to %b?

yes, I think so

**Comment by npostavs** [comment]
As I was going to update the manual, I noticed the `magit-rename-buffer` function which made me realize there is a flaw in this unquifiying scheme: it doesn't distinguish between buffers renamed by uniquify and buffers renamed by the user. So if you call `(magit-status "sub2/repo")` twice (assuming there is already a `"sub1/repo"` open), on the second call you get a new status buffer called `*magit-status: repo<sub2>*<2>`. The problem is present even without uniquify, where you would get a new `*magit-status: repo*<3>` instead of the existing `*magit-status: repo*<2>`.

**Comment by tarsius** [comment]
> As I was going to update the manual, I noticed the `magit-rename-buffer` function which made me realize there is a flaw in this unquifiying scheme:

Just thought about that too yesterday and was going to mention it here :-)

> on the second call you get a new status buffer called `*magit-status: repo<sub2>*<2>`.

First, it doesn't make sense to use `magit-rename-buffer` with the status buffer, it is intended for diff, log, and similar buffers, where you might actually want multiple buffers (per-repo) open at the same time.

Secondly, I occasionally use `magit-rename-buffer` (without the uniquify changes) and never noticed any problems. So you might want to try again with a diff buffer, and without using uniquify. Also I don't quite understand your description of what is going wrong.

**Comment by npostavs** [comment]
> Secondly, I occasionally use magit-rename-buffer (without the uniquify changes) and never noticed any problems. So you might want to try again with a diff buffer, and without using uniquify. Also I don't quite understand your description of what is going wrong.

I omitted some details, let me try again.

The problem occurs when using a `%b` format and having at least 2 buffers in repositories with the same base name open. So if you have `magit-diff-buffer-name-format` = `*magit-diff: %b*`, and you open a diff buffer in `/foo/repo` and then a 2nd diff buffer in `/other-foo/repo`, then the 2nd diff buffer will be named `*magit-diff: repo*<2>`, i.e., Emacs already renames without you calling `magit-rename-buffer`.

 Opening a diff buffer in the 2nd repo again will open yet another `*magit-diff: repo*<3>`.

So this isn't directly related to the uniquify stuff, except that probably nobody would want to use `%b` without uniquify.

**Comment by tarsius** [comment]
I would like to postpone this after v2.2.0, which I would like to release on the 1th August (or if necessary a few days after that).

**Comment by tarsius** [comment]
This will conflict heavily with #2192. I am closing this, but intend to port these changes onto the things I am working on in `jb/get-buffer-by-mode-and-args`, so don't delete this branch just yet.
